### PR TITLE
cart: Compare applied coupon codes when adjusting item qty

### PR DESCRIPTION
**Pull request by ChristopherSchwab**
When adjusting items to their possibly restricted quantity checks if the AppliedCouponCodes of the cart change and provides a template function for getting if this is the case.

### CODE REVIEW

File: cart/application/cartService.go [hunk 1]
```diff
@@ -54,11 +54,12 @@ type (
...
-		DeliveryCode      string
-		WasDeleted        bool
-		RestrictionResult *validation.RestrictionResult
-		NewQty            int
+		OriginalItem          cartDomain.Item
+		DeliveryCode          string
+		WasDeleted            bool
+		RestrictionResult     *validation.RestrictionResult
+		NewQty                int
+		HasRemovedCouponCodes bool
```
Comment (carstendietrich) [inline cart/application/cartService.go]: maybe other naming according to `WasDeleted`

`WereCouponCodesRemoved`
`CouponsCodesWereRemoved`
`RemovedCouponCodes`

File: cart/application/cartService.go [hunk 2]
```diff
@@ -945,13 +946,44 @@ func (cs *CartService) dispatchAllEvents(ctx context.Context, events []flamingo.
...
+			return nil, err
+		}
+		qtyAdjustmentResult.HasRemovedCouponCodes = !couponCodes.ContainedIn(cart.AppliedCouponCodes)
+	}
 
+	return qtyAdjustmentResults, nil
+}
+
+// GenerateRestrictedQtyAdjustments checks the quantity restrictions for each item of the cart and returns which items should be adjusted and how
+func (cs *CartService) GenerateRestrictedQtyAdjustments(ctx context.Context, session *web.Session) (QtyAdjustmentResults, error) {
```
Comment (carstendietrich) [inline cart/application/cartService.go]: Is this a useful exported/public function or only a helper which could be unexported?

File: cart/application/cartService.go [hunk 3]
```diff
@@ -1013,3 +1037,14 @@ func (cs *CartService) getProductWithActiveVariantIfProductIsConfigurable(ctx co
 
 	return product, nil
 }
+
+// HasRemovedCouponCodes returns if a QtyAdjustmentResults has any adjustment that removed a coupon code from the cart
+func (qar QtyAdjustmentResults) HasRemovedCouponCodes() bool {
```
Comment (carstendietrich) [inline cart/application/cartService.go]: If updating the struct property maybe also changing the function name.

File: cart/interfaces/templatefunctions/getItemAdjustment.go [hunk 4]
```diff
@@ -83,8 +86,23 @@ func (gdm *GetQuantityAdjustmentUpdatedItemsMessage) Func(ctx context.Context) i
 	}
 }
 
+// Func defines the GetQuantityAdjustmentCouponCodesRemoved template function
+func (gcd *GetQuantityAdjustmentCouponCodesRemoved) Func(ctx context.Context) interface{} {
```
Comment (carstendietrich) [inline cart/interfaces/templatefunctions/getItemAdjustment.go]: Can we have a small test for the template func?

File: cart/application/cartService.go (hunk 1 above)
Comment (ChristopherSchwab) [reply cart/application/cartService.go]: Personally i feel that the current naming is the closest compared to `WasDeleted.`
`RemovedCouponCodes` is the counterpart to `Deleted`, isnt it?

File: cart/application/cartService.go (hunk 2 above)
Comment (ChristopherSchwab) [reply cart/application/cartService.go]: Unexported the helper function for now.

File: cart/application/cartService.go (hunk 3 above)
Comment (ChristopherSchwab) [reply cart/application/cartService.go]: See above.

File: cart/interfaces/templatefunctions/getItemAdjustment.go (hunk 4 above)
Comment (ChristopherSchwab) [reply cart/interfaces/templatefunctions/getItemAdjustment.go]: Added some tests for the template functions related to automated item adjustments.

**Pull request merged**